staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go

```diff
@@ -29,30 +32,116 @@ import (
 	"k8s.io/apimachinery/pkg/apis/meta/v1beta1/validation"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/apiserver/pkg/endpoints/handlers/negotiation"
 	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
 	endpointsrequest "k8s.io/apiserver/pkg/endpoints/request"
+
+	klog "k8s.io/klog/v2"
 )
 
-// transformObject takes the object as returned by storage and ensures it is in
-// the client's desired form, as well as ensuring any API level fields like self-link
-// are properly set.
-func transformObject(ctx context.Context, obj runtime.Object, opts interface{}, target *schema.GroupVersionKind, scope *RequestScope) (runtime.Object, error) {
+// watchEmbeddedEncoder performs encoding of the embedded object.
+//
+// NOTE: watchEmbeddedEncoder is NOT thread-safe.
+type watchEmbeddedEncoder struct {
+	encoder runtime.Encoder
+
+	ctx context.Context
+
+	// target, if non-nil, configures transformation type.
+	// The other options are ignored if target is nil.
+	target  *schema.GroupVersionKind
+	options interface{}
+	scope   *RequestScope
+
+	// identifier of the encoder, computed lazily
+	identifier runtime.Identifier
+}
+
+func newWatchEmbeddedEncoder(ctx context.Context, encoder runtime.Encoder, target *schema.GroupVersionKind, options interface{}, scope *RequestScope) *watchEmbeddedEncoder {
+	result := &watchEmbeddedEncoder{
+		encoder: encoder,
+		ctx:     ctx,
+		target:  target,
+	}
+	if target != nil && target.Kind == "Table" {
```

Comment:
*metav1.TableOptions is the only options type we expect here, so require a positive check and fail if this is non-nil and not *metav1.TableOptions (or change the signature here to take `options *metav1.TableOptions` and check at the caller)